docs/sql/02-type-system.md

```diff
@@ -0,0 +1,175 @@
+The type system is defined by the list of supported types, type mapping, and type conversion rules.
+
+A type is defined by name, precedence, precision, and scale.
+
+- **Name** is a textual representation of type name
+- **Precision** is the total count of significant digits in the whole number, applicable to **numeric types**
+- **Precedence** is a comparable value that is used for type inference in expressions. A type with a
+  higher value has precedence over a type with a lower value. If two types have the same precedence value, then
+  the type with higher precision has precedence.
+
+**Type family** is a collection of types with the same name, but different precisions. All types
+within a family have the same name and precedence. For example, `INTEGER(11)` and `INTEGER(12)` are two types
+from the same `INTEGER` family.
+
+The scale is not used in Hazelcast Mustang. It is applicable only for `DECIMAL`, `REAL`, and `DOUBLE` types.
+Instead of defining it as a separate value, we just treat these types as types with infinite scale.
+
+## 1 Supported Types
+
+Types supported by the Hazelcast Mustang are listed in Table 1.
+
+`OBJECT` is a Hazelcast-specific type representing an object which doesn't match any other type.
+
+`NULL` is a special type representing a type of `NULL` literal to which a more
+specific type couldn't be assigned. Consider `SELECT NULL FROM t` query, the `NULL`
+literal would have `NULL` type assigned.
+
+[//]: # (TODO : CREATE TYPE command)
+
+*Table 1: Hazelcast Mustang Data Types*
+
+| SQL Type                   | Precedence | Precision |
+|----------------------------|------------|-----------|
+| `NULL`                     | 0          |           |
+| `VARCHAR`                  | 100        |           |
+| `BOOLEAN`                  | 200        | 1         |
+| `TINYINT`                  | 300        | 4         |
+| `SMALLINT`                 | 400        | 7         |
+| `INTEGER`                  | 500        | 11        |
+| `BIGINT`                   | 600        | 20        |
+| `DECIMAL`                  | 700        | Unlimited |
+| `REAL`                     | 800        | Unlimited |
+| `DOUBLE`                   | 900        | Unlimited |
+| `TIME`                     | 1000       |           |
+| `DATE`                     | 1100       |           |
+| `TIMESTAMP`                | 1200       |           |
+| `TIMESTAMP WITH TIME ZONE` | 1300       |           |
+| `OBJECT`                   | 1400       |           |
+
+The type `TIME WITH TIME ZONE` is not supported because of its confusing behavior: daylight-saving rules make it hard to reason
+about time with offset without date part. For this reason, this type is of little use for real applications. The support for this
+type might be added in future releases if we find useful use cases for it.
+
+Structured and UDF data types are not supported at the moment. The support for these types will be added in future releases.
```

Comment:
what do you mean by "UDF data types"? In context of SQL "UDF" = user defined function.

```suggestion
Structured and user-defined data types are not supported at the moment. The support for these types may be added in future releases.
```

Reply:
user defined types support in SQL is experimental: https://docs.hazelcast.com/hazelcast/5.2/sql/user-defined-types